storage/bootstrap/process.go

```diff
@@ -68,26 +71,43 @@ func (b *bootstrapProcess) initTimeRanges(writeStart time.Time) xtime.Ranges {
 }
 
 // bootstrap returns the bootstrapped results for a given shard time ranges.
-func (b *bootstrapProcess) bootstrap(shard uint32, timeRanges xtime.Ranges) (ShardResult, error) {
-	results := NewShardResult(b.opts)
-	for i := 0; i < b.opts.GetMaxRetries(); i++ {
-		res, unfulfilled := b.bootstrapper.Bootstrap(shard, timeRanges)
-		results.AddResult(res)
-		if xtime.IsEmpty(unfulfilled) {
-			return results, nil
+func (b *bootstrapProcess) bootstrap(timeRanges xtime.Ranges, shards []uint32) (Result, error) {
+	result := NewResult()
+
+	it := timeRanges.Iter()
+	for it.Next() {
+		shardsTimeRanges := make(ShardTimeRanges, len(shards))
+
+		window := it.Value()
+
+		r := xtime.NewRanges().AddRange(window)
+		for _, s := range shards {
+			shardsTimeRanges[s] = r
 		}
-		// If there are still unfulfilled time ranges, sleep for a while and retry.
-		time.Sleep(b.opts.GetThrottlePeriod())
+
+		b.log.WithFields(
+			xlog.NewLogField("shardsLen", len(shards)),
+			xlog.NewLogField("from", window.Start.String()),
+			xlog.NewLogField("to", window.End.String()),
+			xlog.NewLogField("length", window.End.Sub(window.Start).String()),
+		).Infof("bootstrapping shards for range")
+
+		res, err := b.bootstrapper.Bootstrap(shardsTimeRanges)
```

Comment:
hm..not sure if we want to bail if there is an error, i'd add whatever we've bootstrapped to `result` and rely on background repair to patch the gaps